### PR TITLE
Fix mesh warp

### LINKED ISSUE #1046: Mesh warp returns mesh with nan's in some cases
I'm not sure what's special about this mesh or particles, but the mesh returned has lots of NaNs:

```
# vtk DataFile Version 4.2
vtk output
ASCII
DATASET POLYDATA
POINTS 13022 float
nan nan nan nan nan nan nan nan nan
nan nan nan nan nan nan nan nan nan
nan nan nan nan nan nan nan nan nan
nan nan nan nan nan nan nan nan nan
nan nan nan nan nan nan nan nan nan
37.314 -3.02839 -19.6524 nan nan nan nan nan nan
nan nan nan nan nan nan nan nan nan
```

Attached are the data and studio xlsx file:

[warp_mesh_problem.zip](https://github.com/SCIInstitute/ShapeWorks/files/6030137/warp_mesh_problem.zip)

### PR DESCRIPTION
**Pull request by akenmorris**
This PR fixes #1046 and #1096.

There were multiple things that needed to be fixed.

First, we have to remove degenerate cells from the mesh, so I apply the Triangle filter followed by vtkCleanPolyData.  Even with the other fixes in this PR, the warp still failed with some femur examples due to the degenerate cells created at the cutting plane boundary.

Second, duplicate particles are removed before generation, marked as bad, and not used for subsequent warps.

Third, new vertices and triangles are added such that particles lie exactly on a vertex mesh.  If a particle already resides on a vertex, then no new vertex or triangles are added for this case.

@sheryjoe , I have created this branch from the 6.0 release branch in case we would like to post a 6.0.1 release with this fix.

I believe this PR will allow us to return to the vanilla LibIGL.

### CODE REVIEW

**Comment by medakk** [comment]
Works on ellipsoid_mesh with 256 particles now

image

File: Studio/src/Application/Data/MeshWarper.cpp [hunk 1]
```diff
@@ -71,17 +103,120 @@ void MeshWarper::check_warp_ready()
...
+      if (p[0] == pt[0] && p[1] == pt[1] && p[2] == pt[2]) {
+        same = true;
+      }
+    }
+
+    if (!same) {
+      int vertex_id = this->reference_mesh_->GetPoints()->InsertNextPoint(pt);
+
+      list->SetNumberOfIds(3);
+      list->SetId(0, cell->GetPointId(0));
```
Comment (medakk) [inline Studio/src/Application/Data/MeshWarper.cpp:161]: is the ordering(back face vs front face) correct here? I think it should be:

```
f0 = [0 1 p]
f1 = [1 2 p]
f2 = [2 0 p]
```
(f is face, p is new point, 0 1 2 and old point indices)

File: Libs/Mesh/MeshUtils.cpp [hunk 2]
```diff
@@ -92,46 +92,44 @@ Eigen::MatrixXi MeshUtils::distilFaceInfo(Mesh mesh){
   return Fref_new;
 }
 
-
-Eigen::MatrixXd MeshUtils::generateWarpMatrix(Eigen::MatrixXd TV , Eigen::MatrixXi TF, Eigen::MatrixXd Vref){
-  
+Eigen::MatrixXd MeshUtils::generateWarpMatrix(Eigen::MatrixXd TV, Eigen::MatrixXi TF,
```
[the remark this answers is not in the document]
Comment (akenmorris) [reply Libs/Mesh/MeshUtils.cpp:95]: No, the slice calls near the end complain.

File: Libs/Mesh/MeshUtils.cpp [hunk 3]
```diff
@@ -159,8 +157,10 @@ Mesh MeshUtils::warpMesh(Eigen::MatrixXd movPts, Eigen::MatrixXd W, Eigen::Matri
   return Mesh(outmesh);
 }
 
-bool MeshUtils::warpMeshes(std::vector< std::string> movingPointpaths, std::vector< std::string> outputMeshPaths, Eigen::MatrixXd W, Eigen::MatrixXi Fref, const int numP){
-  
+bool MeshUtils::warpMeshes(std::vector<std::string> movingPointpaths,
```
[the remark this answers is not in the document]
Comment (akenmorris) [reply Libs/Mesh/MeshUtils.cpp:160]: Yeah, this isn't used anywhere and will probably be dropped.

File: Studio/src/Application/Data/MeshWarper.cpp [hunk 4]
```diff
@@ -71,17 +103,120 @@ void MeshWarper::check_warp_ready()
...
+
+//---------------------------------------------------------------------------
+void MeshWarper::find_good_particles()
+{
+  std::set<int> set;  // initially store in set to avoid duplicates
+  for (int i = 0; i < this->points_.rows(); i++) {
+    for (int j = i + 1; j < this->points_.rows(); j++) {
+      double p1[3]{this->points_(i, 0), this->points_(i, 1), this->points_(i, 2)};
+      double p2[3]{this->points_(j, 0), this->points_(j, 1), this->points_(j, 2)};
+      if (p1[0] == p2[0] && p1[1] == p2[1] && p1[2] == p2[2]) {
```
[the remark this answers is not in the document]
Comment (akenmorris) [reply Studio/src/Application/Data/MeshWarper.cpp:191]: Indeed!

File: Studio/src/Application/Data/MeshWarper.cpp (hunk 1 above)
Comment (akenmorris) [reply Studio/src/Application/Data/MeshWarper.cpp:161]: Yes!  Thank you.  I tossed it in without thinking about it and meant to check it later and forgot.

**Review by medakk — APPROVED** [review]
Looks good! Thanks for fixing this

**Comment by sheryjoe** [comment]
>@sheryjoe , I have created this branch from the 6.0 release branch in case we would like to post a 6.0.1 release with this fix.
@akenmorris let's post a 6.0.1 for this fix. Please announce on the discussion forum.

**Pull request merged**